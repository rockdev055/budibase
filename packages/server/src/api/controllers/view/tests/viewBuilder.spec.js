const statsViewTemplate = require("../viewBuilder");

describe("viewBuilder", () => {

  describe("Filter", () => {
    it("creates a view with multiple filters and conjunctions", () => {
      expect(statsViewTemplate({
        "name": "Test View",
        "modelId": "14f1c4e94d6a47b682ce89d35d4c78b0",
        "filters": [{
          "value": "Test",
          "condition": "EQUALS",
          "key": "Name"
        }, {
          "value": "Value",
          "condition": "MT",
          "key": "Yes",
          "conjunction": "OR"
        }]
      })).toMatchSnapshot()
    })
  })

  describe("Calculate", () => {
      expect(statsViewTemplate({
        "name": "Calculate View",
        "field": "myField",
        "modelId": "14f1c4e94d6a47b682ce89d35d4c78b0",
        "filters": []
      })).toMatchSnapshot()
  })

  describe("Group By", () => {
    it("creates a view emitting the group by field", () => {
      expect(statsViewTemplate({
        "name": "Test Scores Grouped By Age",
        "modelId": "14f1c4e94d6a47b682ce89d35d4c78b0",
        "groupBy": "age",
        "field": "score",
        "filters": [], 
      })).toMatchSnapshot()
    })
  })
});