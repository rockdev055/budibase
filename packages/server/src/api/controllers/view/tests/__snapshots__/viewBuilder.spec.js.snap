// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewBuilder Filter creates a view with multiple filters and conjunctions 1`] = `
Object {
  "map": "function (doc) {
      if (doc.modelId === \\"14f1c4e94d6a47b682ce89d35d4c78b0\\" && doc[\\"Name\\"] === \\"Test\\" || doc[\\"Yes\\"] > \\"Value\\") {
        emit(doc._id);
      }
    }",
  "meta": Object {
    "field": undefined,
    "filters": Array [
      Object {
        "condition": "EQUALS",
        "key": "Name",
        "value": "Test",
      },
      Object {
        "condition": "MT",
        "conjunction": "OR",
        "key": "Yes",
        "value": "Value",
      },
    ],
    "groupBy": undefined,
    "modelId": "14f1c4e94d6a47b682ce89d35d4c78b0",
    "schema": Object {
      "avg": "number",
      "count": "number",
      "max": "number",
      "min": "number",
      "sum": "number",
      "sumsqr": "number",
    },
  },
  "reduce": "_stats",
}
`;

exports[`viewBuilder Group By creates a view emitting the group by field 1`] = `
Object {
  "map": "function (doc) {
      if (doc.modelId === \\"14f1c4e94d6a47b682ce89d35d4c78b0\\" ) {
        emit(doc[\\"age\\"], doc[\\"score\\"]);
      }
    }",
  "meta": Object {
    "field": "score",
    "filters": Array [],
    "groupBy": "age",
    "modelId": "14f1c4e94d6a47b682ce89d35d4c78b0",
    "schema": Object {
      "avg": "number",
      "count": "number",
      "max": "number",
      "min": "number",
      "sum": "number",
      "sumsqr": "number",
    },
  },
  "reduce": "_stats",
}
`;
